File: backend/app/schemas/assistant.py
"""
Data validation and serialization schemas (Pydantic v2) for the AI Assistant.

Defines validation models for conversation sessions, chat messages,
RAG queries, and structured source citations returned by the AI Assistant.
"""

import uuid
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field

from app.core.config import settings


class ConversationRole(str, Enum):
    """
    Supported participant roles inside a conversation.
    """

    USER = "user"
    ASSISTANT = "assistant"


class SourceCitation(BaseModel):
    """
    Structured metadata describing a document chunk used during RAG retrieval.
    """

    work_item_id: uuid.UUID = Field(
        ...,
        description="Primary key UUID of the source Work Item.",
    )

    original_filename: str = Field(
        ...,
        description="Original uploaded filename.",
    )

    chunk_index: int = Field(
        ...,
        ge=0,
        description="Zero-based chunk index.",
    )

    similarity_score: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Vector similarity score.",
    )

    page_number: int | None = Field(
        None,
        ge=1,
        description="Optional source page number.",
    )

    snippet: str = Field(
        ...,
        min_length=1,
        description="Retrieved document snippet.",
    )

    model_config = {
        "from_attributes": True
    }


class TokenUsage(BaseModel):
    """
    Standardized token usage metadata returned by every LLM provider.
    """

    provider: str
    model: str

    prompt_tokens: int
    completion_tokens: int
    total_tokens: int

    estimated_cost: float


class ConversationMessageBase(BaseModel):
    """
    Base schema shared across conversation message models.
    """

    role: ConversationRole = Field(
        ...,
        description="Conversation participant role.",
    )

    content: str = Field(
        ...,
        min_length=1,
        description="Message content.",
    )

    sources: list[SourceCitation] | None = Field(
        None,
        description="Structured RAG citations associated with this message.",
    )

    token_usage: TokenUsage | None = Field(
        None,
        description="Token usage metadata for assistant messages.",
    )


class ConversationMessageCreate(BaseModel):
    """
    Request payload used when submitting a new user message.
    """

    content: str = Field(
        ...,
        min_length=1,
        description="User message.",
    )


class ConversationMessageResponse(ConversationMessageBase):
    """
    Response schema representing a stored conversation message.
    """

    id: uuid.UUID
    conversation_id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True
    }


class ConversationBase(BaseModel):
    """
    Base schema shared across conversation models.
    """

    title: str = Field(
        ...,
        min_length=1,
        max_length=settings.MAX_CONVERSATION_TITLE_LENGTH,
        description="Conversation title.",
    )

    work_item_id: uuid.UUID | None = Field(
        None,
        description="Optional associated Work Item.",
    )


class ConversationCreate(BaseModel):
    """
    Request payload used when creating a new conversation.
    """

    title: str | None = Field(
        None,
        max_length=settings.MAX_CONVERSATION_TITLE_LENGTH,
        description="Optional custom title. Defaults to the first user message.",
    )

    work_item_id: uuid.UUID | None = Field(
        None,
        description="Optional Work Item for document assistant mode.",
    )


class ConversationUpdate(BaseModel):
    """
    Request payload used when renaming an existing conversation.
    """

    title: str = Field(
        ...,
        min_length=1,
        max_length=settings.MAX_CONVERSATION_TITLE_LENGTH,
        description="Updated conversation title.",
    )


class ConversationResponse(ConversationBase):
    """
    Response schema representing a conversation session.
    """

    id: uuid.UUID
    user_id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    messages: list[ConversationMessageResponse] = Field(
        default_factory=list,
        description="Chronological conversation messages.",
    )

    model_config = {
        "from_attributes": True
    }


class ChatQuery(BaseModel):
    """
    Request payload submitted to the AI Assistant.
    """

    content: str = Field(
        ...,
        min_length=1,
        description="User prompt or instruction.",
    )


class ChatResponse(BaseModel):

    response: str = Field(
        ...,
        description="Generated assistant response.",
    )

    sources: list[SourceCitation] = Field(
        default_factory=list,
        description="Document citations supporting the generated response.",
    )

    token_usage: TokenUsage = Field(
        ...,
        description="Token usage metadata for the generated response.",
    )

    model_config = {
        "from_attributes": True
    }

File: frontend/src/types/assistant.ts
/**
 * AI Assistant and Conversational Memory Data Transfer Objects (DTOs)
 * for FlowPilot AI.
 *
 * Defines immutable TypeScript contracts shared between the frontend
 * and backend while preparing the application for streaming responses,
 * conversation pagination, telemetry collection, and future multi-agent
 * orchestration.
 */

/* ========================================================================== */
/* Common Types                                                               */
/* ========================================================================== */

/**
 * Supported conversation participant roles.
 */
export type ConversationRole =
  | "user"
  | "assistant";

/**
 * Reserved agent categories for future multi-agent workflows.
 */
export type AgentType =
  | "assistant"
  | "rag"
  | "workflow"
  | "classification"
  | "extraction"
  | "system";

/**
 * Supported model identifiers.
 *
 * Extend this union whenever additional providers/models are introduced.
 */
export type AIModelIdentifier =
  | "gpt-5"
  | "gpt-4.1"
  | "claude"
  | "gemini"
  | "llama"
  | "mistral"
  | "custom";

/**
 * Conversation lifecycle state.
 */
export type ConversationStatus =
  | "ACTIVE"
  | "ARCHIVED"
  | "DELETED";

/* ========================================================================== */
/* Citation Models                                                            */
/* ========================================================================== */

/**
 * Metadata describing a document chunk referenced during RAG generation.
 */
export interface SourceCitation {
  /**
   * Stable unique identifier.
   * React rendering should never depend on array indexes.
   */
  readonly citation_id: string;

  readonly work_item_id: string;

  /**
   * Original uploaded filename.
   */
  readonly original_filename: string;

  /**
   * Optional UI-friendly display name.
   * Allows future renaming without affecting storage.
   */
  readonly document_display_name?: string;

  readonly chunk_index: number;

  readonly similarity_score: number;

  readonly page_number: number | null;

  readonly snippet: string;
}

/* ========================================================================== */
/* Message Models                                                             */
/* ========================================================================== */

/**
 * Single chronological message belonging to a conversation.
 */
export interface ConversationMessage {
  readonly id: string;

  readonly conversation_id: string;

  /**
   * Guarantees deterministic ordering during streaming.
   */
  readonly sequence_number: number;

  readonly role: ConversationRole;

  readonly content: string;

  readonly sources: readonly SourceCitation[];

  readonly created_at: string;

  readonly updated_at: string;

  /* ---------------------------------------------------------------------- */
  /* Future Multi-Agent Metadata                                             */
  /* ---------------------------------------------------------------------- */

  readonly agent_name?: string;

  readonly agent_type?: AgentType;

  readonly model?: AIModelIdentifier;

  readonly token_usage?: TokenUsage;
}

/* ========================================================================== */
/* Conversation Models                                                        */
/* ========================================================================== */

/**
 * Lightweight conversation metadata.
 *
 * Used when listing conversations without loading
 * the complete message history.
 */
export interface ConversationSummary {
  readonly id: string;

  readonly user_id: string;

  readonly title: string;

  readonly work_item_id: string | null;

  readonly status: ConversationStatus;

  readonly created_at: string;

  readonly updated_at: string;
}

/**
 * Cursor-paginated message history.
 *
 * Designed for future infinite scrolling.
 */
export interface ConversationHistoryResponse {
  readonly messages: readonly ConversationMessage[];

  readonly total_messages: number;

  readonly has_more: boolean;

  readonly next_cursor: string | null;
}

/* ========================================================================== */
/* Request DTOs                                                               */
/* ========================================================================== */

/**
 * User chat request payload.
 */
export interface ChatQueryRequest {
  readonly content: string;
}

/**
 * Create conversation request.
 *
 * Title generation is delegated to the backend
 * using the user's first message.
 */
export interface ConversationCreateRequest {
  readonly work_item_id?: string | null;
}

/**
 * Manual conversation rename request.
 */
export interface ConversationUpdateRequest {
  readonly title: string;
}

export interface ConversationHistoryQuery {
  readonly limit?: number;
  readonly cursor?: string | null;
}

/* ========================================================================== */
/* Telemetry                                                                  */
/* ========================================================================== */

/**
 * Token accounting returned by the LLM provider.
 */
export interface TokenUsage {
  readonly provider: string;

  readonly model: string;

  readonly prompt_tokens: number;

  readonly completion_tokens: number;

  readonly total_tokens: number;

  readonly estimated_cost: number;
}

/**
 * Optional performance telemetry.
 *
 * Designed for future analytics dashboards.
 */
export interface ChatTelemetry {
  readonly usage?: TokenUsage;

  readonly latency_ms?: number;

  readonly model?: AIModelIdentifier;
}

/* ========================================================================== */
/* Response DTOs                                                              */
/* ========================================================================== */

/**
 * Assistant response returned after a chat request.
 *
 * Compatible with traditional request/response APIs
 * and future streaming implementations.
 */
export interface ChatResponse {
  /**
   * Conversation identifier.
   *
   * Useful when the backend automatically creates
   * a new conversation.
   */
  readonly conversation_id: string;

  /**
   * Assistant message text.
   */
  readonly response: string;

  /**
   * Referenced document chunks.
   */
  readonly sources: readonly SourceCitation[];

  /**
   * Indicates whether the streamed response
   * has completed.
   *
   * Optional for non-streaming APIs.
   */
  readonly is_complete?: boolean;

  /**
   * Optional performance metrics.
   */
  readonly telemetry?: ChatTelemetry;
}
